### PR TITLE
array-match: specify how many times GetIterator is called

### PR DESCRIPTION
**Pull request by zkat**
Fixes: #99

### CODE REVIEW

File: CORE.md [hunk 1]
```diff
@@ -249,6 +249,11 @@ _ObjectMatchPattern_ `:` `{` _MatchProperty_ `}`
 
 1. TODO
 
+_ArrayMatchPattern_ `:` TKTK
+
+1. Let _iter_ be ? GetIterator(_exprValue_)
+2. ...only run GetIterator once per value-position in destructuring... (grouping together multiple check against the same ArrayMatchPattern such that info is shared across matches, but only for specific positions (that is, given patterns `[a, b]`, `[...foo]`, and `{x: [...foo]}`, GetIterator will be called once: once for the top-level arraypatterns and once for the nested one -- _even if_ the iterable in the third pattern is object-identical to the toplevel patterns))
```
Comment (bathos) [inline CORE.md]: should this say "GetIterator will be called twice: once for ... and once for"?

Comment (zkat) [reply CORE.md]: I just re-read this and it definitely reads the way I intended.

I'll expand and reword this to make it clearer.

Comment (bathos) [reply CORE.md]: @ljharb the "even if" makes it sound like the first two share one iteration result while a nested occurrence of the same object is considered distinct and will need to be iterated again (which  makes sense: you cannot know statically that the nested reference is circular, so it cannot contribute to determining the max length needed for a shared iterator). But I’m not sure I’m understanding right, since it also says it is only called once overall.

Comment (zkat) [reply CORE.md]: @bathos no, you pretty much understood right. "by value-position" means the two toplevel ones share a single iteration of the iterator, and the nested one calls it twice -- even if the nested one and the toplevel one are object-identical:

```js
const arr = [1,2,3]
arr.x = arr
match (arr) {
  [a,b] ~> ... // GetIterator()
  [...c] ~> ... // Reuses that iterator
  {x: [...d]} ~> ... // GetIterator() again, even though arr.x === arr
}
```

make sense?

Comment (ljharb) [reply CORE.md]: Note that it’s not multiple GetIterator calls that are the issue imo, it’s that the result of every next call needs to be compared to every match, not discarded - because in many iterables, each GetIterator call will return the same (already advanced) iterator.
